### PR TITLE
Authentication

### LINKED ISSUE #35: Session-based authentication
# Description
After this task, users should be able to register, login, and use our app securely. No more hardcoding `const coopId = 1;` :)

Specifically,
- Passwords should now be salted and hashed in the `coops` table
- Register creates a new entry in the `coops` table
- Login checks if username and (hashed) password match and generates a session token
- Redirect to Login if we try to access FiltersPage/MyProfile without being logged in
- Redicted to FiltersPage if we try to access Login/Register while being logged in

# Suggested steps:
1. `yarn add` the following dependencies
- bcrypt
- express-session
- passport
- passport-local
2. Read through the **Overview**, **Authenticate**, and **Configure** sections of the [Passport.js documentation](http://www.passportjs.org/docs/). By now, you should be comfortable navigating through complex documentation. Reach out if anything still confuses you.
3. Create a file `passport-config.js` that sets up the the initial middleware and `passport-local` [username & password](http://www.passportjs.org/docs/username-password/) strategy. Make sure to use PostgreSQL syntax (`db.query`) rather than the Mongoose syntax in their examples. Be sure to `require` this on `app.js` and add `passport` to `module.exports` for it to actually take effect.
4. Modify the `CREATE TABLE coop` command in `db/cofed.sql` to add a new column `salt`
5. Add a `POST /auth/register` endpoint to `api/routes/auth.js` that
- checks to ensure the email does not currently exist in the database (sending back an error if it does)
- generates a salt and hashes the password using `bcrypt`
- stores username, co-op name, salt, hashed password, and location
6. Implement the `verify` callback that computes the hashed password and uses `bcrypt.compare` to compare to our result in the database. Also make sure `serializeUser` and `deserializeUser` is implemented.
7. Add a `POST /auth/login` endpoint to `api/routes/auth.js`. Hook up LoginPage to use this endpoint.
8. Define a new [middleware](https://expressjs.com/en/guide/using-middleware.html) function (has the parameters `req, res, next`) and add it to all of the existing GET and POST handlers on `api/routes/index.js`! See [this](https://scotch.io/tutorials/route-middleware-to-check-if-a-user-is-authenticated-in-node-js) if you're still confused. Ensure that FiltersPage and ProfilePage are only accessible if the user is authenticated; otherwise redirect to LoginPage.
9. Similarly, define another middleware function so that the user is redirected to FiltersPage when trying to access LoginPage or RegisterPage while logged in.
10. Implement logout (see the Operations section in the Passport documentation!)

# (Potentially) Helpful Resources
- [Passport](http://www.passportjs.org/docs/) documentation
- [bcrypt](https://www.npmjs.com/package/bcrypt) documentation
- Quick [100-second overview](https://www.youtube.com/watch?v=UBUNrFtufWo) of user authentication (ignore JWT and tokens)
- Guy with fancy British accent explains [cookies](https://www.youtube.com/watch?v=LHSSY8QNvew) (only the first ~4 minutes or so is relevant to us)

### PR DESCRIPTION
**Pull request by zaidmamsa1**
## Motivation

Motivation: high. Maybe 1 week left!

Fixes #35, session-based authentication

## Summary of Changes

Feel free to use sentences or a bulleted list.

- Create new session for user upon correct login information input
- Ensured authentication for profile and filterpage
- Hashed password and compared to non-hashed input
- Ended user session upon logout
- Authenticated and redirected accordingly to register page

## Testing

N/A. No need 😎

### CODE REVIEW

File: api/routes/auth.js [hunk 1]
```diff
@@ -1,6 +1,10 @@
 const express = require('express');
 const db = require('../../db/index');
 const router = express.Router();
+const bcrypt = require('bcrypt');
+const saltFactor = 10;
```
Comment (leebianca) [inline api/routes/auth.js]: nit: by convention, constant identifiers are in uppercase so just change to `const SALT_FACTOR = 10`

File: api/app.js [hunk 2]
```diff
@@ -11,12 +12,35 @@ const port = 8000;
 app.use(express.json());
 app.use(express.urlencoded({ extended: true }));
 
-// Routes - may need to add more
-app.use('/api', indexRouter);
-app.use('/auth', authRouter);
+// // Routes - may need to add more
+// app.use('/api', indexRouter);
+// app.use('/auth', authRouter);
```
Comment (rrrliu) [inline api/app.js]: remove

File: api/app.js [hunk 3]
```diff
@@ -11,12 +12,35 @@ const port = 8000;
...
+
+//   // CHECK THE USER STORED IN SESSION FOR A CUSTOM VARIABLE
+//   // you can do this however you want with whatever variables you set up
+//   if (req.user.authenticated) {
+//     return next();
+//   } else {
+//     res.status(404).send({ error: `user not logged in` });
+//   }
+//   // IF A USER ISN'T LOGGED IN, THEN REDIRECT THEM SOMEWHERE
+// }
```
Comment (rrrliu) [inline api/app.js]: remove

File: api/passport-config.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        WHERE email = $1;`,
+      [username]
+    );
+    //result is the password from the DATABASE
+    if (query.rows.length > 0) {
+      //get the query result
+      const user = query.rows[0];
+      //compare the password in the DATABASE with the ENTERED password
+      // you should be hashing it when you're putting it INTO the database, not when you're pulling it out. so you might need to modify your register logic -richard
+      // in fact, just delete the INSERT INTO lines in cofed.sql and make your own users
```
Comment (rrrliu) [inline api/passport-config.js]: remove

File: api/passport-config.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+      let hashpass = user.hashed_pass; //await bcrypt.hash(user.hashed_pass, saltFactor);
+      const passwordMatch = await bcrypt.compare(password, hashpass);
+      //if successful return the user
+      if (passwordMatch) {
+        console.log('===== MATCHED PASSWORD =====');
+        return done(null, user.id);
+      } else {
+        console.log('PASSWORD = ' + password);
+        console.log('INCORRECT PASS: ' + user.hashed_pass);
```
Comment (rrrliu) [inline api/passport-config.js]: pls remove

File: api/passport-config.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        WHERE id = $1;`,
+      [id]
+    );
+    if (query.rows.length > 0) {
+      return done(null, query.rows[0]);
+    } else {
+      return done(null, null);
+    }
+  } catch (err) {
+    return done(err, null);
```
Comment (rrrliu) [inline api/passport-config.js]: I think it’s just done(err)

File: api/routes/auth.js [hunk 7]
```diff
@@ -14,4 +18,48 @@ router.get('/', async (req, res) => {
...
+  try {
+    const { email, name, addr, password } = req.body;
+    const email_query = await db.query(
+      `SELECT id FROM coops WHERE email = $1;`,
+      [email]
+    );
+    //checks to ensure the email does not currently exist in the database
+    if (email_query.rows.length == 0) {
+      //generate salt
+      // let salt = await bcrypt.genSalt(saltFactor);
```
Comment (rrrliu) [inline api/routes/auth.js]: rm

File: client/src/App.js [hunk 8]
```diff
@@ -33,7 +33,7 @@ function App() {
         <Route exact path="/register" component={RegisterPage} />
         <Route
           exact
-          path="/createprofile"
+          path="/profilepage"
```
Comment (rrrliu) [inline client/src/App.js]: actually, we can just delete this route because we’re not using it

File: api/passport-config.js [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+      [username]
+    );
+    //result is the password from the DATABASE
+    if (query.rows.length > 0) {
+      //get the query result
+      const user = query.rows[0];
+      //compare the password in the DATABASE with the ENTERED password
+      // you should be hashing it when you're putting it INTO the database, not when you're pulling it out. so you might need to modify your register logic -richard
+      // in fact, just delete the INSERT INTO lines in cofed.sql and make your own users
+      let hashpass = user.hashed_pass; //await bcrypt.hash(user.hashed_pass, saltFactor);
```
Comment (leebianca) [inline api/passport-config.js]: remove comment from this line

**Review by leebianca — CHANGES_REQUESTED** [review]
OMG YAY proud of you guys for figuring out authentication! just left a few minor comments here and there but for the most part, looks GOOD

File: api/routes/auth.js [hunk 10]
```diff
@@ -1,17 +1,46 @@
...
+require('../passport-config');
 
-// This is an example SQL query - use this as a template
-router.get('/', async (req, res) => {
+router.post('/register', async (req, res) => {
   try {
-    const query = await db.query(
-      `SELECT 'This is the root of authRouter.' as text;`
+    const { email, name, addr, password } = req.body;
+    const email_query = await db.query(
```
Comment (rrrliu) [inline api/routes/auth.js]: nit: camelCase please

File: api/routes/index.js [hunk 11]
```diff
@@ -102,21 +109,21 @@ router.delete('/delete', async (req, res) => {
   }
 });
 
-router.post('/coop', async (req, res) => {
-  const { email, name, addr, pass } = req.body;
+router.post('/coop', isAuthenticated, async (req, res) => {
+  const { email, name, addr, hashed_pass } = req.body;
```
Comment (rrrliu) [inline api/routes/index.js]: nit: camelCase hashed_pass -> hashedPassword (make sure to change on frontend as well)

File: client/src/components/Navbar/Navbar.js [hunk 12]
```diff
@@ -1,34 +1,47 @@
...
+      setRedirect(true);
+    } catch (err) {
+      console.log(err.stack);
+      setRedirect(false);
+    }
+  }
+
+  if (redirect) {
+    return <Redirect to="/login" />;
+  } else {
```
Comment (rrrliu) [inline client/src/components/Navbar/Navbar.js]: nit: remove else

File: client/src/pages/LoginPage/LoginPage.js [hunk 13]
```diff
@@ -1,93 +1,100 @@
...
-      console.log('error!');
+  async function verify() {
+    try {
+      await axios.post('/auth/login', {
+        username: emailInput,
+        password: pwInput,
+      });
+      setRedirect(true);
+    } catch (err) {
+      console.log(err.stack);
```
Comment (rrrliu) [inline client/src/pages/LoginPage/LoginPage.js]: rm

File: client/src/pages/ProfilePage/ProfilePage.js [hunk 14]
```diff
@@ -14,12 +27,21 @@ export default function ProfilePage() {
...
   React.useEffect(() => {
     async function fetchData() {
-      const res = await axios.get('/api/coop/' + CoopId);
-      setProfileVariables(res.data);
+      try {
+        const res = await axios.get('/api/coop');
+        setProfileVariables(res.data);
+      } catch (err) {
+        setRedirect(true);
+        console.log(err.stack);
```
Comment (rrrliu) [inline client/src/pages/ProfilePage/ProfilePage.js]: rm

File: client/src/pages/RegisterPage/RegisterPage.js [hunk 15]
```diff
@@ -8,66 +8,76 @@ export default function RegisterPage() {
...
-        <p className="inputType">Co-op Name</p>
-        <input
-          className="register-input-box"
-          type="text"
-          value={nameInput}
-          onChange={e => setNameInput(e.target.value)}
-        />
+  if (redirect) {
+    return <Redirect to="/login" />;
+  } else {
```
Comment (rrrliu) [inline client/src/pages/RegisterPage/RegisterPage.js]: nit: rm else

File: client/src/components/Navbar/Navbar.js [hunk 16]
```diff
@@ -1,34 +1,47 @@
...
-      </Link>
-      <div className="navbar-rightside">
+  const [redirect, setRedirect] = React.useState(false);
+
+  async function logout() {
+    try {
+      await axios.post('/auth/logout');
+      setRedirect(true);
+    } catch (err) {
+      console.log(err.stack);
```
Comment (rrrliu) [inline client/src/components/Navbar/Navbar.js]: rm

File: client/src/pages/FiltersPage/FiltersPage.js [hunk 17]
```diff
@@ -21,23 +22,28 @@ export default function FiltersPage() {
...
-    const res = await axios.get('/api/coops');
-    setCoops(res.data);
-    setCoopShown(res.data[0]);
+    try {
+      const res = await axios.get('/api/coops');
+      setCoops(res.data);
+      setCoopShown(res.data[0]);
+    } catch (err) {
+      setRedirect(true);
+      console.log(err.stack);
```
Comment (rrrliu) [inline client/src/pages/FiltersPage/FiltersPage.js]: rm

**Review by rrrliu — CHANGES_REQUESTED** [review]

**Review by rrrliu — DISMISSED** [review]

File: db/cofed.sql [hunk 18]
```diff
@@ -50,33 +50,19 @@ GRANT ALL PRIVILEGES ON ALL SEQUENCES IN SCHEMA public TO root;
...
 
-INSERT INTO stars
-    (starred_coop_id, starrer_coop_id)
+INSERT INTO coop_tags
+    (coop_id, tag_id)
 VALUES
+    (1, 1),
     (1, 2),
     (1, 3),
     (2, 1),
```
Comment (rrrliu) [inline db/cofed.sql]: we don't have any coops, so delete these

File: client/src/pages/FiltersPage/FiltersPage.js [hunk 19]
```diff
@@ -37,14 +39,16 @@ export default function FiltersPage() {
   }
 
   async function fetchInitialData() {
-    const res = await axios.get('/api/coops');
-    setCoops(res.data);
-    if (res.data.length !== 0) {
```
Comment (rrrliu) [inline client/src/pages/FiltersPage/FiltersPage.js:42]: keep this `if (res.data.length !== 0)` condition

**Review by rrrliu — CHANGES_REQUESTED** [review]
last few fixes

**Review by rrrliu — APPROVED** [review]

**Review by leebianca — APPROVED** [review]

**Pull request merged**